### ISSUE TITLE
Add error handling to the FPM setup banner

### ISSUE THREAD

**Issue by techanvil** [google/site-kit-wp#9846]
## Feature Description

The initial implementation of the FPM setup banner doesn't handle the case where the call to save the settings fails. The banner is simply unconditionally dismissed:

https://github.com/google/site-kit-wp/blob/bf1bec6e07f7172f4fda5222efc5306e17fccbcd/assets/js/components/notifications/FirstPartyModeSetupBanner.js#L63-L68

We should handle an error here in a similar way to the Consent Mode setup CTA which will display the error in the banner:

https://github.com/google/site-kit-wp/blob/bf1bec6e07f7172f4fda5222efc5306e17fccbcd/assets/js/components/consent-mode/ConsentModeSetupCTAWidget.js#L274-L278

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* In the FPM setup banner, if an error is returned by the call to save the FPM settings:
  * The error should be displayed above the CTA buttons.
  * The banner should not be dismissed.
  * The success banner should not be displayed.

## Implementation Brief

**Note:** 

1. In this implementation we will refactor the notifications API such that we can introduce some reusable error handling in notification components. Also, we will improve a bit of code that relates to the notification dismissals.

2. There's a POC PR #9878 created by @techanvil  which can be referenced during the implementation.

- [x] In `assets/js/googlesitekit/notifications/components/common/CTALink.js`
  - Accept the following props in the component:
    - `dismissOnCTAClick`: Boolean to determine whether or not notification should get dismissed on CTA click if there is no error.
    - `dismissOptions`: It is useful in order to change the dismiss behaviour when CTA is clicked.
    - `dismissExpires`: It will be part of `dismissOptions` and denotes the expiration time for dismissal after which the notification must reappear. Zero means the permanent dismissal.
  - We shall use `clearError` and `receiveError` actions from `core/site` store for error management within the component.
  - In `handleCTAClick` function, clear the errors using `clearError` action as soon as CTA is clicked.
  - Await for `onCTAClick` to receive any that might be returned by the function. The relevant changes are mentioned below `onCTAClick` to return error.
  - If the error is returned by `onCTAClick`, set it using `receiveError`. Set `baseName` for error as `notificationAction` and pass the notification id `id` as argument.
  - If `dismissOnCTAClick` is `true`, add `dismissNotification` function `ctaClickActions` using push method and pass `dismissOptions` to it, along with `expiresInSeconds` equal to `dismissExpires`.
  - Write an effect which should clear any errors related to CTA when the component is unmounted.

- [x] In `assets/js/googlesitekit/notifications/components/common/ActionsCTALinkDismiss.js`
  - Accept the following props in the component:
    - `ctaDismissOptions`: This is further passed down to `CTALink` component as `dismissOptions` prop. It is useful in order to change the dismiss behaviour when CTA is clicked.
    - `dismissOnCTAClick`: Boolean to determine whether or not notification should get dismissed on CTA click if there is no error.
  - Also, pass `dismissOptions` and `dismissOnCTAClick` props to `CTALink` component.

- [x] Create a component `Error` in `assets/js/googlesitekit/notifications/components/common`
  - It will receive the notification id as the prop to retrieve the error for concerned notfication.
  - Retrieve any errors for `notificationAction` set in `CTALink` component above. If the error is present, return `ErrorText` component with error message, else return null.
  - Add an effect to clear the error when component unmounts. Refer POC.

- [x] In `assets/js/components/notifications/FirstPartyModeSetupBanner.js`
  - Change `onCTAClick` function to be `async`. Await for `saveFirstPartyModeSettings` action to be completed and collect the `error`returned by the action.
  - If `error` is present, return the `error` object as it is.
  - Remove the call for `dismissNotification` action as we want dismiss behaviour to happen only within notification components itself and within CTA handlers.
  - Pass `ctaDismissOptions` to `ActionsCTALinkDismiss` component with the value set to object as `{skipHidingFromQueue: false}`

- In `assets/js/googlesitekit/notifications/components/layout/NotificationWithSVG.js`
  - Render `Error` component just above the `actions`. Pass the notification id to the component.

### Test Coverage

1. Update the tests for `FirstPartyModeSetupBanner` component such that the new tests for testing error behaviours will be added.

2. Add a new scenario for error state in the `FirstPartyModeSetupBanner` story.

3. Add tests for `Error` component.

## QA Brief

1. Install [tweak extension](https://tweak-extension.com/docs/intro) and add following rule. Note that `domain.com` should be replaced with your site domain.

  * URL: http://domain.com/wp-json/google-site-kit/v1/core/site/data/fpm-settings?_locale=user
  * HTTP Method: POST
  * Status code: 500
  * Delay: 2000
  * Response payload:
  ```
     {
      "code": "test_error",
      "message": "Error message.",
      "data": { "reason": "" }
    }
  ```

2. Setup the FPM banner on Site Kit dashboard. Refer #9647 for the steps to setup the FPM banner.

3. Refresh the Site Kit dashboard. Enable tweak extension with the rule above on SK dahboard.

4. Click on CTA, error message should appear above the CTA.

## Changelog entry

* Add error handling to the First-party mode setup banner to display errors and prevent premature dismissal when saving settings fail.

**Comment by techanvil** [comment]
Hi @ankitrox, thanks for drafting the IB. However, rather than handling this with logic specific to the `FirstPartyModeSetupBanner` component, we have a good opportunity to provide some reusable error handling to the generic notification components.

At the same time, we can improve a piece of code that relates to the notification dismissals.

I've created a PoC PR for my suggestion: https://github.com/google/site-kit-wp/pull/9878

I'd like to get @jimmymadon's eyes on this too, as it relates to the banner notifications refactor.

Note that we could further extend the error handling to support an error being returned by the `onDismiss()` handler, as well as the `onCTAClick()` which is fleshed out in the PR.

https://github.com/google/site-kit-wp/blob/faac928ab9aaf2a702c50fa3cdafeda166df8cde/assets/js/googlesitekit/notifications/components/common/Dismiss.js#L49-L59

**Comment by jimmymadon** [comment]
@techanvil I've added my comments - overall, the reusability in the POC looks great. Thanks for cooking this up.

>Note that we could further extend the error handling to support an error being returned by the onDismiss() handler,

Yes, absolutely.

**Comment by techanvil** [comment]
Hi @ankitrox, thanks for reworking the IB. I have, however reviewed @jimmymadon's feedback on the PoC and pushed an update in response to https://github.com/google/site-kit-wp/pull/9878#discussion_r1892576754. See https://github.com/google/site-kit-wp/pull/9878#discussion_r1892576754.

Please can you update the IB accordingly?

Btw, you don't need to go into so much detail in the IB with the PoC PR to refer to, you can provide a higher level overview and refer to the PR for the detail. The IB should of course highlight any identified variations from the PoC that should be made if there are any.

**Comment by techanvil** [comment]
Note that I've created an issue, https://github.com/google/site-kit-wp/issues/9925, to follow on from the work here, adding error handling for banner dismissals and extending error handling across all of the refactored banner notifications.

**Comment by ankitrox** [comment]
Thanks @techanvil. I've updated the IB as per latest changes.

**Comment by techanvil** [comment]
Thanks @ankitrox, that LGTM. The estimate of 19 seems a bit on the high side though, considering the logic is fleshed out in the PR, so most of the work will be around testing. Would you agree and if so revisit the estimate?

**Comment by ankitrox** [comment]
Yes I agree @techanvil . Sorry it was not updated because it was set when we did not have the POC PR ready. I think the estimate of 11 will be appropriate as most of the implementation work is already present in the POC and we mainly need to focus on tests.

**Comment by techanvil** [comment]
Thanks @ankitrox! That SGTM.

IB ✅

**Comment by mohitwp** [comment]
### QA Update ✅ 

- Tested on dev environment.
- Verified that in the FPM setup banner, if an error is returned by the call to save the FPM settings:

       --  The error should be displayed above the CTA buttons. - **PASS**
       --  The banner should not be dismissed. - **PASS**
       --  The success banner should not be displayed. - **PASS**

- Tested using extension to mock error.
- Verified that error appeared above the CTA and banner is not getting dismiss.

<details>

https://github.com/user-attachments/assets/a878a715-e509-42a0-9ed8-6d573ef47f40

![Image](https://github.com/user-attachments/assets/88edfb5d-0218-49fc-a1d3-f36ef69ae644)

![Image](https://github.com/user-attachments/assets/6960b3eb-6961-473c-81ce-e52a503af4cb)

</details>
